app/services/topbar_sweeper.rb

```diff
@@ -0,0 +1,22 @@
+class TopbarSweeper
+  include Singleton
+  attr_accessor :controller
+
+  def expire_cache(user = User.current)
+    controller.expire_fragment(self.class.fragment_name(user.id)) if controller.present? && user.present?
+  end
+
+  class << self
+    delegate :expire_cache, to: :instance
+
+    def fragment_name(id = User.current.id)
+      "tabs_and_title_records-#{id}"
+    end
+
+    def expire_cache_all_users
+      User.unscoped.pluck(:id).each do |id|
+        Rails.cache.delete("views/#{fragment_name(id)}")
+      end
+    end
+  end
+end
```

Comment:
I'm not sure that I understand this comment. Broken down:

> and it's thread unfriendly singleton with unnecessary reference to controller

One thread can only be processing one request at a time, I'm not sure that thread-safety is a requirement in controllers or helpers like this.

> Method expire_fragment which is only available in controller context should be really called from proper place and not from here. [..] I think it can be all done properly in a controller helper.

Do you want this implemented as a controller callback on any action that modifies any of the affected models? That could be a large number of actions. Keeping the callback on the model (and then this singleton to link the model callback to the calling controller) is likely to ensure consistency better, no matter where a model is changed from.